src/dmd/backend/dlist.d

```diff
@@ -0,0 +1,570 @@
+/**
+ * Interface to the C linked list type.
+ *
+ * List is a complete package of functions to deal with singly linked
+ * lists of pointers or integers.
+ * Features:
+ *      1. Uses mem package.
+ *      2. Has loop-back tests.
+ *      3. Each item in the list can have multiple predecessors, enabling
+ *         different lists to 'share' a common tail.
+ *
+ * Copyright:   Copyright (C) 1986-1990 by Northwest Software
+ *              Copyright (c) 1999-2016 by Digital Mars, All Rights Reserved
+ * Authors:     $(LINK2 http://www.digitalmars.com, Walter Bright)
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(DMDSRC backend/tk/_dlist.d)
+ */
+
+module dmd.backend.dlist;
+
+import core.stdc.stdarg;
+import core.stdc.stdio;
+import core.stdc.stdlib;
+import core.stdc.string;
+
+extern (C++):
+
+nothrow @nogc
+{
+
+/* **************** TYPEDEFS AND DEFINES ****************** */
+
+struct LIST
+{
+        /* Do not access items in this struct directly, use the         */
+        /* functions designed for that purpose.                         */
+        LIST* next;             /* next element in list                 */
+        int count;              /* when 0, element may be deleted       */
+        union
+        {       void* ptr;      /* data pointer                         */
+                int data;
+        }
+}
+
+alias LIST* list_t;             /* pointer to a list entry              */
+
+/* FPNULL is a null function pointer designed to be an argument to
+ * list_free().
+ */
+
+alias void function(void*) @nogc nothrow list_free_fp;
+
+enum FPNULL = cast(list_free_fp)null;
```

Comment:
You can avoid the cast with `enum list_free_fp FPNULL = null`. But defining a new symbol here doesn't really buy a lot in comparison to just using `null`.

Reply:
FPNULL dates from the olden days when function pointers were different from regular pointers due to the wacky DOS memory models. It doesn't make sense any more, as you point out, but changing it here violates the #1 rule of translation.